src/navigation/reroute.js

```diff
@@ -109,6 +109,16 @@ export function reroute(pendingPromises = [], eventArguments) {
         )
       );
 
+      if (typeof navigationIsCanceled === "function") {
+        try {
+          navigationIsCanceled = await navigationIsCanceled();
+        } catch (e) {
+          throw new Error(
+            "CancelNavigation if passed a function, the function must return a promise and the promise must resolve to either true or false"
```

Comment:
This error should be improved in the following ways:

1. It should be optimized out of the production bundle. See https://github.com/single-spa/single-spa/blob/14aa441e609bd5b29374ab2dd0706c789d71992c/src/applications/apps.js#L106-L111 as an example
2. The `e` variable should not be swallowed, as that error is important for people diagnosing what happened. Swallowing it makes it impossible for people to know what happened without putting a breakpoint on this line of code.
3. The error should be assigned a number and then given a corresponding explanation page in the single-spa.js.org documentation. We've used error numbers 1 to 41, so this one could take error 42. Here's where you add an error code documentation page: https://github.com/single-spa/single-spa.js.org/tree/master/website/src/components/Errors/Codes
4. The current code hard fails the reroute when the user-provided navigationIsCanceled function throws an error, which prevents single-spa applications from mounting/unmounting. I don't think that's the best behavior - instead I think it would make sense to assume the navigation is not canceled when the navigationIsCanceled function throws an error. That way the system continues functioning somewhat despite bad code provided by the user. To do that, I think we could create a new function in app-errors.js called `handleGeneralError` that calls setTimeout and then calls all error handlers (or throws the error if there are no error handlers).